### DISCUSSION TITLE
Why FastAPI redirects url with a trailing slash instead of resolving it internally?

### DISCUSSION THREAD

**Discussion by tusharsnx** [fastapi/fastapi#9328] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()
api_users_router = APIRouter(prefix="/api/users")

@api_users_router.get("/")
async def users_index():
    return { status: "success" }

app.include_router(api_users_router)
```

### Description

FastAPI tends to prefer redirecting urls with a trailing slash, e.g. `/path/to/resource` to `/path/to/resource/` when that's possible.

```python
from fastapi import FastAPI

api_users_router = APIRouter(prefix="/api/users")

@router.get("/")
async def api_users_index():
    return { status: "success" }

app = FastAPI()
app.include_router(api_users_router)
````

The above snippet would redirect any call to `/api/users` to `/api/users/` causing another full round trip.

I found this opposite of what happens on other API libraries for e.g. Node's Expressjs. It resolves both URLs with/without trailing slash to the same route handler without having to configure anything else. 

The current behaviour has several drawbacks when used behind proxy servers with TLS:
- The redirecting response actually redirects to `http://` instead for `https://`.
- Some HTTP clients E.g. `aiohttp` drop the `POST` (maybe other methods aswell) request due to the change in security protocol level (https to http) and instead make a `GET` request in this case. This all happens silently and generates unexpected results.
- Instead of one redirect, it actually causes 2 redirects (or total 3 round trips). First, FastAPI sends RedirectResponse with `http://` then, proxy server redirecting back to `https://` and finally browser makes the request with correct scheme and url. For Browsers, this might not be an issue since 2nd redirect happens internally due to HSTS, but for http clients (eg. axios/nodejs, aiohttp/python) these requests introduce latency and bad UX.

I want to make a PR in this regard which could utilize an option like resolve_without_slash:

```python
from fastapi import FastAPI

api_users_router = APIRouter(prefix="/api/users", resolve_without_slash=True)

@api_users_router.get("/")
async def api_users_index():
    return { status: "success" }

app = FastAPI()
app.include_router(api_users_router)
````

which should be equivalent to:
```python
from fastapi import FastAPI

api_users_router = APIRouter(prefix="/api/users")

@api_users_router.get("")
@api_users_router.get("/")
async def api_users_index():
    return { status: "success" }


app = FastAPI()
app.include_router(api_users_router)
````

Maybe I'm being silly here and making wrong assumptions but I'm interested to know if the new behaviour would be any better than what it is now.

### Operating System

Linux

### FastAPI Version

0.89.1

### Python Version

Python 3.11.2

**Comment by ivarflakstad** [comment]
I don't know if this discussion answers your question as to _why_, but it does provide some solutions to the problem.
https://github.com/tiangolo/fastapi/discussions/7298

**Reply by tusharsnx** [reply]
I already went through all the workaround in that thread and even used the double decoration of methods once with `/` and then without it.

But I'm trying to get an explanation for why this isn't done automatically? We can easily choose to opt-out of it if we need the alternate behavior i.e redirect.

**Reply by ivarflakstad** [reply]
Yeah, I suspected it wasn't satisfactory hehe.
At the very least I agree with the decision of separating between trailing slash or not as a default.
[To slash or not to slash](https://developers.google.com/search/blog/2010/04/to-slash-or-not-to-slash).
But why there isn't a convenient way to override that default I do not know.

One small hurdle I can think of (iiuc) is that this behaviour is specified per router, so if you want to provide a flag for overriding the behaviour globally it would have to be propagated to all routers.

As yet another alternative to what has been discussed in the other thread I _think_ you can achieve this by looping through all the routes in your app after it's initialised and add a duplicate route with/without trailing slash if it doesn't exist.

You could probably also create a middleware that tries to complete the request against both alternatives.

**Comment by pjaydev** [comment]
Just a little note here, which does not fully regard to your suggestion, but to the following statement: 

> * The redirecting response actually redirects to `http://` instead for `https://`.

I had the same problem and nearly spent half a day about finding the solution. In my researches I also came across this thread and want to add a comment in case it might help others with a similar problem.
 
The reason that the scheme of the redirect URL changes could be due to the fact that the original client request is not completely forwarded to the Uvicorn server or not properly processed by it.

When operating FastAPI behind an HTTPS proxy, the following should be ensured so that the original request URL can be correctly reconstructed on the server side:

1. The `X-Forwarded-For` and `X-Forwarded-Proto` headers should be set correctly by the proxy server. The following configuration must be set when using nginx for example:

```
location ... {
	...
		proxy_set_header Host $http_host;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header X-Forwarded-Proto $scheme;
		....
		proxy_pass http://yourbackend;
	....	
}
```

2. In order for Uvicorn to take these headers in account when reconstructing the request URL, uvicorn must be started with the following flags.

	`uvicorn --proxy-headers --forwarded-allow-ips=*`

  The last parameter `forwarder-allow-ips` defines, from which remote IPs the `X-Forwarded...` headers are accepted and is particularly important if uvicorn and nginx are operated on different servers or in different Docker containers and therefore have different local IP addresses, since `forwarded-allow-ips` would otherwise default to `127.0.0.1`.

When following these steps, redirection URLs (and the URL provided via `Request.url`) should contain the proper `https://` prefix.

**Reply by brianoflondon** [reply]
Just to add a BIG thanks for this. In my case the addition of `--forwarded-allow-ips=*` to the CMD in my Dockerfile was all I needed. I had been hunting around in Traefik configs which was a red herring.

```
# This was added with refrence to this link: https://github.com/tiangolo/fastapi/discussions/9328#discussioncomment-8242245
CMD ["gunicorn", "v4vapp_api_ext.main:app", "--proxy-protocol", "--forwarded-allow-ips=*", "--workers", "4", "--worker-class", "uvicorn.workers.UvicornWorker", "--bind", "0.0.0.0:80"]
```

And it would appear that the Traefik proxy I'm using automatically adds the correct headers without need for any further config.

**Reply by adsouza** [reply]
That is technically a security hole.

**Comment by Synrom** [comment]
Hi everyone! 🎉 I wanted to share an update related to this discussion. I’ve submitted a pull request that introduces a new boolean flag, `ignore_trailing_slash`, to the FastAPI class. This flag, when set to True, allows FastAPI to handle URIs with and without trailing slashes equivalently, without causing a 307 redirect.

Using the flag is as easy as:
```python
app = FastAPI(ignore_trailing_slash=True)
```

The pull request is already approved by a maintainer, but still awaits feedback from Tiangolo himself, so community interest or input in general would be greatly appreciated. If this feature would be helpful for you, feel free to check out the pull request and share your thoughts or support.

**Reply by tusharsnx** [reply]
This is a great news. Thanks 😃

**Reply by peteski22** [reply]
Would this change how the OpenAPI docs are generated? 

I ask this because (personally) I'd like to see them generated with the shortest required URI to hit the API. 

e.g. `/api/users` shown in the docs vs. `/api/users/`

It also means the docs look more uniform against other endpoints you may have added such as:

```python
@api_users_router.get("/{user_id}")
async def user_by_id():
```

Using the example from the original post...

Preferred:
![image](https://github.com/user-attachments/assets/ce8fa6ec-d440-4490-a248-604c76133a86)


Existing (when using `@api_users_router.get("/")`):
![image](https://github.com/user-attachments/assets/20755a56-a453-4d1c-b412-4e1cbf276df6)

But would be good to understand if there's any change, or optional change. 

Thanks!

**Reply by peteski22** [reply]
Just in case anyone else comes along looking, you can find @Synrom's [PR here](https://github.com/fastapi/fastapi/pull/12145).

**Reply by Synrom** [reply]
Hey @peteski22,

thank you for the feedback.
Yes it does change how the docs are generated and the trailing slash does not appear in the docs when `ignore_trailing_slash` is set :)

**Reply by slburson** [reply]
This would be great — would have saved me several hours today.  NextJS, by default, issues redirects to strip trailing slashes, and then FastAPI will redirect again to put the slash back on if it's needed.  If NextJS is being used as a reverse proxy, as in my case, this isn't even correct, because FastAPI sends an absolute URL (including host and port) back with the redirect.  (Maybe it should send a relative URL?)

I would almost suggest that `ignore_trailing_slash` should be true by default, but that's probably too risky.

**Reply by gaby** [reply]
@Synrom I was excited to try this only to realize that the PR still open after 7 months 😂

**Comment by asmaier** [comment]
Maybe this is of interest here:

Since [FastAPI version 0.98.0](https://fastapi.tiangolo.com/pt/release-notes/#0980) the framework provides a way to disable the redirect behaviour by setting the `redirect_slashes` parameter to `False`, which is `True` by default.

See
- https://stackoverflow.com/a/77902129/179014

**Comment by tiangolo** [comment, marked as answer]
From what I understand, the root problem is the redirect interactions with a TLS Termination Proxy and FastAPI redirecting automatically to a non HTTPS URL.

I realized I didn't support the flag in FastAPI CLI to properly set that up, so I just tweaked and merged a PR implementing it, and released `fastapi-cli` version 0.0.10.

Then, what I think is the most important part, I documented how all this works, the right configuration, how the TLS Termination Proxy passes the headers to FastAPI with the current URL: https://fastapi.tiangolo.com/advanced/behind-a-proxy/

In short, if your app is behind a trusted proxy and all requests go through that trusted proxy, you can use:

```bash
fastapi run --forwarded-allow-ips="*"
```

That will tell the application server to trust all the _forwarded_ headers coming from all IPs (assuming that the only client is the proxy), so that it can extract those headers and get the correct host and use it when creating the redirects.
